File: C-Plus-Plus/dp/dijsktra.cpp
#include<bits/stdc++.h>
using namespace std;
//Function to find the minimum vertex
int findMinVertex(int* distance, bool visited[], int n) {
	//Initialize min value
	int minVertex = -1;
	for (int i = 0; i < n; i++) {
		if (!visited[i] && (minVertex == -1 || distance[i] < distance[minVertex])) {
			minVertex = i;
		}
	}
	return minVertex;
}

//Dijkstra function
void dijkstra(int** edges, int n) {
	//hold the shortest distance from src to i
	int* distance = new int[n];
	//check if it is visited or not?
	bool* visited = new bool[n];
	for (int i = 0; i < n; i++) {
		//Initialize all distances as INFINITE and visited[] as false
		distance[i] = INT_MAX;
		visited[i] = false;
	}
	distance[0] = 0;
	//Now pick all corrosponding vertex and explore them
	for (int i = 0; i < n - 1; i++) {
		//finding min vertex i,e minimum then source
		int minVertex = findMinVertex(distance, visited, n);
		visited[minVertex] = true;
		//Now check its Neighbour
		for (int j = 0; j < n; j++ ) {
			if (edges[minVertex][j] != 0 && !visited[j]) {
				int dist = distance[minVertex] + edges[minVertex][j];
				if (dist < distance[j]) {
					distance[j] = dist;
				}
			}
		}
	}
	//print the constructed distance array
	cout << "Vertex\tDistance from Source\n";
	for (int i = 0; i < n - 1; i++) {
		cout << i << "          " << distance[i] << endl;
	}
}

//Main function
int main() {
	//Take input from user
	cout << "Enter value for node and edge -";
	int n, e;
	cin >> n >> e;
	int** edges = new int*[n];
	for (int i = 0; i < n; i++) {
		edges[i] = new int[n];
		for (int j = 0; j < n; j++) {
			edges[i][j] = 0;
		}
	}
	cout << "Enter value for Adjacency Matrix -";
	for (int i = 0; i < e; i++) {
		int f, s, weight;
		cin >> f >> s >> weight;
		edges[f][s] = weight;
		edges[s][f] = weight;
	}
	//Calling  Disjkstra function
	cout << endl;
	dijkstra(edges, n);
}

/*
Sample Input Output

Enter value for node and edge -
5 6
Enter value for Adjacency Matrix -
0 1 4
0 2 8
1 3 5
1 2 2
2 3 5
2 4 9

Vertex	Distance from Source
0          0
1          4
2          6
3          9

Time Complexity -O(N^2)
*/

File: Java/Greedy_Algorithms/Dijsktra.java
/*
/* Problem Description- This is a shortest path algorithm used in graphs which is used to find the shortest 
distance from a starting vertex to all the vertex. The edge between two vertices having minimum distance is 
considered and also no vertex is left alone and not connected.
 */

 import java.util.*;
 import java.lang.*;
 class dijsktra
 {
     static int Ver;
     int minDistance(int dist[], Boolean sptSet[])
     {
         // Minimum value intialized with maximum value
         int min = Integer.MAX_VALUE;
         int  min_index = -1; //setting minimum index to -1 as array index starts from zero.

         for (int v = 0; v < Ver; v++)
             if (sptSet[v] == false && dist[v] <= min) {
                 min = dist[v];
                 min_index = v;
             }

         return min_index;
     }

     // A utility function to print the constructed distance array
     void printSolution(int dist[])
     {
         System.out.println("Vertex \t  Distance from Source");
		 for (int i = 0; i < Ver; i++)
             System.out.println(i + " \t\t " + dist[i]);
	 }
	 //to implement dijkstra algorithm by finding the shortest distance
     void dijkstra(int graph[][], int src)
     {
         int dist[] = new int[Ver]; //This will have the shortest distance
         // sptSet[i] will true if vertex i is included in shortest
         Boolean sptSet[] = new Boolean[Ver];
         for (int i = 0; i < Ver; i++) {
             dist[i] = Integer.MAX_VALUE;//all distance initalized as infinite
             sptSet[i] = false;
         }
         dist[src] = 0; //since the sorce vertex will always have distance 0 from itself/

         // Find shortest path for all vertices
         for (int count = 0; count < Ver - 1; count++) {
             // Pick the minimum distance vertex from the set of vertices
             int u = minDistance(dist, sptSet);
             // Mark the picked vertex as processed
             sptSet[u] = true;
             for (int v = 0; v < Ver; v++)

                 // Update dist[v] only if is not in sptSet, there is an
                 // edge from u to v, and total weight of path from src to
                 // v through u is smaller than current value of dist[v]
                 if (!sptSet[v] && graph[u][v] != 0 && dist[u] != Integer.MAX_VALUE && dist[u] + graph[u][v] < dist[v])
                     dist[v] = dist[u] + graph[u][v]; //updating the distance
         }
         printSolution(dist);//calling the print function to print the distance form vertex.
     }
     public static void main(String[] args) //main class
     {
         Scanner sc=new Scanner(System.in);
         System.out.println("Enter number of vertices");
         Ver=sc.nextInt();
         int graph[][] = new int[Ver][Ver];
         System.out.println("Enter graph");
         for(int i=0;i<Ver;i++)
         {
             for(int j=0;j<Ver;j++)
             {
                 graph[i][j]=sc.nextInt();
				}
         } 
         dijkstra t = new dijkstra();
         t.dijkstra(graph, 0);
    }
 }


 /* 
 Input/Output -
 Enter number of vertices 3 
 Enter graph
 0
 0
 2
 0
 0
 0
 6
 7 
 0
Vertex  Distance from Source 
  0        0
  1        9
  2        2 

 Time Complexity: O(Ver)
 Space Complexity: O(Ver)
  */
